### PR TITLE
dataloader, midi tensors (.py and .pynb)

### PR DESCRIPTION
**Pull request by Madhudorai**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,13 +8,7 @@ https://colab.research.google.com/drive/1ykem3MaFS2XqVXz7ZniUxtqXQL0vqpK4?usp=sh
...
 
 # Collab Notebook 
-Open the .ipynb in google colab. It automatically downloads the dataset from Zenodo URL and uses pytorch dataloaders to load the dataset. You can filter the dataset through 
-a) players= 'P1','P2','P3' and combinations of these
-b) content types = 'scales', 'chords','music', 'techniques', 'singlenotes' 
-c) modalities = 'directinput', 'micamp', 'exo', 'ego'
-
-Labels of note, note onset, note offset, fret, string are generated using the MIDI file, we can decide the time slice of the sample (using slice range or slice size)- so that all modalities and the corresponding labels are aligned for each sample. 
-We can hear the different modalities and look at the labels for each sample. 
+Open the .ipynb in google colab. It automatically downloads the dataset from Zenodo URL and uses pytorch dataset to load the dataset. We can also use a dataloader for batches.
```
Comment (iranroman) [inline README.md]: `uses a pytorch dataset to load the data. One can also pass this to a pytorch dataloader to create batches for model training.`

File: README.md [hunk 2]
```diff
@@ -8,13 +8,7 @@ https://colab.research.google.com/drive/1ykem3MaFS2XqVXz7ZniUxtqXQL0vqpK4?usp=sh
...
-Open the .ipynb in google colab. It automatically downloads the dataset from Zenodo URL and uses pytorch dataloaders to load the dataset. You can filter the dataset through 
-a) players= 'P1','P2','P3' and combinations of these
-b) content types = 'scales', 'chords','music', 'techniques', 'singlenotes' 
-c) modalities = 'directinput', 'micamp', 'exo', 'ego'
-
-Labels of note, note onset, note offset, fret, string are generated using the MIDI file, we can decide the time slice of the sample (using slice range or slice size)- so that all modalities and the corresponding labels are aligned for each sample. 
-We can hear the different modalities and look at the labels for each sample. 
+Open the .ipynb in google colab. It automatically downloads the dataset from Zenodo URL and uses pytorch dataset to load the dataset. We can also use a dataloader for batches.  
 
 # Citations:
```
Comment (iranroman) [inline README.md:13]: also include the bibtex for the paper please.

File: guitartechs_dataset/dataset.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+                                if prefix in ['Set1', 'Set2', 'Set3', 'Set4']:
+                                    chord_type = '3-note chord'
+                                elif prefix == 'Drop3':
+                                    chord_type = '4-note chord'
+                            self.index.append({
+                                'player': player,
+                                'content_type': content,
+                                'sample': sample_value,
+                                'chord_type': chord_type
+                            })
```
Comment (iranroman) [inline guitartechs_dataset/dataset.py]: let's turn this into a method please

Comment (iranroman) [inline guitartechs_dataset/dataset.py]: I'm still not convinced that this is the best representation for the dataset's index. This is because I find it a bit awkward that we are being specific about chord types but not about anything else. Also, we will carry the `chord_type` key even when things may not be chords. 

I understand why this way it "works" given the dataset's metadata, and perhaps we can leave this as is for now, but keep in mind for future revisions.

File: guitartechs_dataset/dataset.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+              # Use librosa utils for slicing
+              frames = librosa.util.frame(y_padded, frame_length=slice_samples, hop_length=hop_length)
+
+              # For each frame, compute start and end time (in seconds)
+              for s in range(frames.shape[1]):
+                  start_sample = s * hop_length
+                  start_sec = start_sample / self.sr
+                  end_sec = start_sec + self.slice_dur
+                  self.expanded_index.append((i, start_sec, end_sec))
```
Comment (iranroman) [inline guitartechs_dataset/dataset.py]: all of this is more complicated than what it has to be. You should not need to load the audio here and compute these slices and paddings "by hand". 

Librosa has more utilities that allow you to compute these slices. Make sure you study and consider using these: https://librosa.org/doc/0.11.0/util.html#array-operations

File: guitartechs_dataset/dataset.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+                        if fname.startswith('directinput_') and fname.endswith('.wav'):
+                            # The sample identifier is based on the file name.
+                            sample_value = fname.replace('directinput_', '').replace('.wav', '')
+                            chord_type = None
+                            if content.lower() == 'chords':
+                                prefix = sample_value.split('_')[0]
+                                if prefix in ['Set1', 'Set2', 'Set3', 'Set4']:
+                                    chord_type = '3-note chord'
+                                elif prefix == 'Drop3':
+                                    chord_type = '4-note chord'
```
Comment (iranroman) [inline guitartechs_dataset/dataset.py]: let's work with the original chord names. Set1, Set2, Drop3, etc. actually mean specific types of chords among the guitar player community. Look it up if interested or of not I'm also happy to explain later.

**Comment by iranroman** [comment]
Hi @Madhudorai 

Thanks. This is looking very good. I left some comments to streamline and improve some parts of the code. Many things are currently more complicated than what they have to be. That said, I did check and the code seems to be functional, which is great. This means we have a full template to work with now. 

Let's also include a simple visualization to check that the onset of sounds and events in the labels tensor are correct.

@vyau please also have a look and if you want to start pushing changes to this same PR, you totally should!

File: guitartechs_dataset/dataset.py (hunk 5 above)
Comment (Madhudorai) [reply guitartechs_dataset/dataset.py]: yes, removed the chord type! makes sense not to be specific about chords alone (like you said seems off). Fixed in [Commit 1d1fbf7](https://github.com/Madhudorai/Guitar-TECHS/commit/1d1fbf7caf70f53b0b5d069667fd049ec7ca8587)

File: guitartechs_dataset/dataset.py (hunk 3 above)
Comment (Madhudorai) [reply guitartechs_dataset/dataset.py]: merged vyau's PR for the same

File: guitartechs_dataset/dataset.py (hunk 4 above)
Comment (Madhudorai) [reply guitartechs_dataset/dataset.py:113]: the paddings are by hand. 
librosa.util.fix_length(data, *, size, axis=-1, **kwargs)[[source]](https://librosa.org/doc/0.11.0/_modules/librosa/util/utils.html#fix_length)[](https://librosa.org/doc/0.11.0/generated/librosa.util.fix_length.html#librosa.util.fix_length)
Fix the length an array data to exactly size along a target axis.
maybe i can replace it with this function? 

slicing is using librosa.util.frame. 

the computing of start and end time is just for the metadata.

Comment (Madhudorai) [reply guitartechs_dataset/dataset.py]: I think its fixed in [Commit f971875](https://github.com/Madhudorai/Guitar-TECHS/commit/f9718751fb31979527dc4d15fff31e4ffb992af3)? do check it out

File: guitartechs_dataset/dataset.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        it is padded with zeros at the end.
+        """
+        start_sample = int(start * self.sr)
+        # Determine desired slice length in samples.
+        desired_length = int(self.slice_dur * self.sr) if self.slice_dur else int((end - start) * self.sr)
+        end_sample = start_sample + desired_length
+        segment = audio[start_sample: min(len(audio), end_sample)]
+        if len(segment) < desired_length:
+            segment = np.pad(segment, (0, desired_length - len(segment)), mode='constant')
+        return segment
```
Comment (iranroman) [inline guitartechs_dataset/dataset.py:228]: A lot of the code in these methods should be unnecessary if one uses the librosa.load parameters of offset and duration.

**Comment by iranroman** [comment]
Thanks @Madhudorai  This is looking pretty good. I would just like to see some sanity checks.

**Comment by iranroman** [comment]
> Let's also include a simple visualization to check that the onset of sounds and events in the labels tensor are correct.

This is what I mean. @vyau could you take the lead on this?
